include/region_map.h

```diff
@@ -1,142 +1,1392 @@
+				    [&](typename types::unique_inner_node_ptr& in) { insert_subtree(o.first, std::move(in)); }); //
+			}
+
+			if(!m_root->contains_leaves() && m_root->num_children() == 1) {
+				// Decrease tree height by 1 level.
+				m_root = std::move(m_root->eject_only_child());
+				m_root->set_depth(0);
+			}
+		}
 
 		/**
-		 * Merge regions with the same values.
+		 * Calculates whether two boxes can be merged. In order to be mergeable, the two boxes
+		 * have to touch in one dimension and match exactly in all remaining dimensions.
 		 */
-		void collapse_regions() {
-			std::set<size_t> erase_indices;
-			for(auto i = 0u; i < m_region_values.size(); ++i) {
-				const auto& values_i = m_region_values[i].second;
-				for(auto j = i + 1; j < m_region_values.size(); ++j) {
-					const auto& values_j = m_region_values[j].second;
-					if(values_i == values_j) {
-						m_region_values[i].first = GridRegion<3>::merge(m_region_values[i].first, m_region_values[j].first);
-						erase_indices.insert(j);
+		bool can_merge(const GridBox<Dims>& box_a, const GridBox<Dims>& box_b) const {
+			bool adjacent = false;
+			for(size_t d = 0; d < Dims; ++d) {
+				if(box_a.get_min()[d] != box_b.get_min()[d] || box_a.get_max()[d] != box_b.get_max()[d]) {
+					// Dimension does not match exactly, but could still be adjacent.
+					// If we already are adjacent in another dimension, we cannot merge.
+					if(!adjacent && (box_a.get_max()[d] == box_b.get_min()[d] || box_b.get_max()[d] == box_a.get_min()[d])) {
+						adjacent = true;
+					} else {
+						return false;
 					}
 				}
 			}
 
-			for(auto it = erase_indices.rbegin(); it != erase_indices.rend(); ++it) {
-				m_region_values.erase(m_region_values.begin() + *it);
+			assert(adjacent);
+			return true;
+		}
+
+		/**
+		 * Try to merge a list of candidate entries with their neighbors within the tree.
+		 */
+		void try_merge(std::vector<typename types::entry>&& merge_candidates) {
+#if !defined(NDEBUG)
+			// Sanity check: Merge candidates do not overlap
+			GridRegion<Dims> candidate_union;
+			for(auto& [box, value] : merge_candidates) {
+				assert(GridRegion<Dims>::intersect(candidate_union, box).empty());
+				candidate_union = GridRegion<Dims>::merge(candidate_union, box);
+			}
+#endif
+
+			// For each candidate, probe around it in every direction to check whether there is a box with same value.
+			//   If yes, check if it can be merged.
+			//     If yes, erase the two boxes, insert the new one and add it as a merge candidate.
+			// Repeat until no more merges are possible.
+			bool did_merge = true;
+			std::vector<bool> merged(merge_candidates.size(), false);
+			while(did_merge) {
+				did_merge = false;
+
+				for(size_t i = 0; i < merge_candidates.size(); ++i) {
+					if(merged[i]) continue;
+					auto& [box, value] = merge_candidates[i];
+
+					// TODO PERF: Order of dimensions can affect merge results
```

Comment:
I think the current first dimension tried is already our preferred split dimension, right? That intuitively seems like it would make sense.

Reply:
Starting with dim0 means that we'll prefer larger columns over rows in 2D (assuming 0=y, 1=x), i.e., striding the faster dimension (1). I *think* doing it the other way round *could* be potentially be better in some cases, but we'd have to do some real world experiments first. This is the way I had it for CCGrid so I decided to leave it for now.